### PR TITLE
support use function pointer in curated template query

### PR DESCRIPTION
**Pull request by YannanGao-gs**
#### What type of PR is this?

Improvement

#### What does this PR do / why is it needed ?

support use function pointer in curated template query

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java [hunk 1]
```diff
@@ -217,11 +227,67 @@ else if (executable instanceof DataSpaceTemplateExecutable)
...
+                                        {
+                                            Optional<? extends ValueSpecification> fromFunc = ((Root_meta_pure_metamodel_function_ConcreteFunctionDefinition_Impl<?>) templateExecutableQuery)._expressionSequence().toList().stream()
+                                                            .filter(func -> func instanceof Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl && ((Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl) func)._functionName.equals("from")).findAny();
+                                            if (fromFunc.isPresent())
+                                            {
+                                                Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl fromFuncExpression = (Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl) fromFunc.get();
+
+                                                // only check mapping and runtime if using ->form() in the most basic way e.g. ->from(model::Mapping, model::Runtime)
```
Comment (travisstebbins) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java:248]: Nit: small typo
```suggestion
                                                // only check mapping and runtime if using ->from() in the most basic way e.g. ->from(model::Mapping, model::Runtime)
```

File: legend-engine-xts-data-space/legend-engine-xt-data-space-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/DataSpaceParseTreeWalker.java [hunk 2]
```diff
@@ -257,9 +257,21 @@ private DataSpaceExecutable visitDataSpaceTemplateExecutable(DataSpaceParserGram
...
+            executable.query = new PackageableElementPointer(
+                    PackageableElementType.FUNCTION,
+                    functionContext.functionIdentifier().getText(),
+                    walkerSourceInformation.getSourceInformation(functionContext.functionIdentifier())
+            );
+        }
+        else if (queryContext.combinedExpression() != null)
+        {
+           executable.query = visitLambda(queryContext.combinedExpression());
+        }
```
Comment (travisstebbins) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/DataSpaceParseTreeWalker.java:274]: What happens if `queryContext.functionIdentifier()` and `queryContext.combinedExpression()` are both null?

Comment (YannanGao-gs) [reply legend-engine-xts-data-space/legend-engine-xt-data-space-grammar/src/main/java/org/finos/legend/engine/language/pure/grammar/from/DataSpaceParseTreeWalker.java:274]: Thanks! I throw an exception now for unsupported query type

File: legend-engine-xts-data-space/legend-engine-xt-data-space-pure-metamodel/src/main/resources/core_data_space_metamodel/metamodel.pure [hunk 3]
```diff
@@ -71,7 +71,7 @@ Class meta::pure::metamodel::dataSpace::DataSpacePackageableElementExecutable ex
 Class meta::pure::metamodel::dataSpace::DataSpaceTemplateExecutable extends meta::pure::metamodel::dataSpace::DataSpaceExecutable
 {
    id: String[1];
-   query: FunctionDefinition<Any>[1];
+   query: Any[1];
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-pure-metamodel/src/main/resources/core_data_space_metamodel/metamodel.pure:74]: As discussed, lets avoid using the type `Any` here. 

Maybe we can have an abstract template class here, with inline and pointer sub classes, with default implementation going to inline to make it backward compatible.

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/test/java/org/finos/legend/engine/language/pure/compiler/test/TestDataSpaceCompilationFromGrammar.java [hunk 4]
```diff
@@ -988,6 +988,143 @@ public void testDataSpaceWithTemplateExecutable()
...
+                "  ];\n" +
+                "  defaultExecutionContext: 'Context 1';\n" +
+                "  elements: [model::element, model, model::sub];\n" +
+                "  executables:\n" +
+                "  [\n" +
+                "    {\n" +
+                "      id: 1;\n" +
+                "      title: 'Template 1';\n" +
+                "      query: model::templateFunc2():TabularDataSet[1];\n" +
+                "      executionContextKey: 'Context 1';\n" +
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/test/java/org/finos/legend/engine/language/pure/compiler/test/TestDataSpaceCompilationFromGrammar.java:1122]: Lets think about what we want to support here. 
Should we support standalone TDS functions without the `->from()` with mapping/runtime and just apply the from from the executionContext provided ? 
If they do provide `->from()` then your check below makes sense but then they should not provide an executoinContextkey and we should throw if they do ?

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java [hunk 5]
```diff
@@ -282,6 +310,71 @@ else if (dataSpace.supportInfo instanceof DataSpaceSupportCombinedInfo)
                         }
                         metamodel._supportInfo(supportInfo);
                     }
+                },
+                (dataSpace, context) ->
+                {
+                    // Move checking mapping/runtime utilized in the function within the curated template query does not align with the mapping/runtime applied in the execution context
+                    // to the sixth pass because function body and parameter are processed in the fifthPassBuilder
+                    // if we do the checking in the fifth pass, function body and parameter will be an empty list
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java:318]: Seems we moved processing of functions from the second pass to the fifth pass in order to parallelize. 
Here: https://github.com/finos/legend-engine/pull/2918
Which is why we started failing the tests. 
Lets mention that PR in this comment.

File: legend-engine-xts-data-space/legend-engine-xt-data-space-protocol/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/dataSpace/DataSpaceTemplateExecutable.java [hunk 6]
```diff
@@ -14,11 +14,9 @@
 
 package org.finos.legend.engine.protocol.pure.v1.model.packageableElement.dataSpace;
 
-import org.finos.legend.engine.protocol.pure.v1.model.valueSpecification.raw.Lambda;
-
 public class DataSpaceTemplateExecutable extends DataSpaceExecutable
 {
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-protocol/src/main/java/org/finos/legend/engine/protocol/pure/v1/model/packageableElement/dataSpace/DataSpaceTemplateExecutable.java:18]: Could we instead do something like this. Don't like the idea combining classes like Lambda/PackageablePointer. 
For the below we can make the inline class the default implementation so it is backward compatible. 
````
public abstract class DataSpaceTemplateExecutable extends DataSpaceExecutable
{
    public String id;
    public String executionContextKey;
}

public class DataSpaceInlineTemplateExecutable extends DataSpaceTemplateExecutable 
{
    public Lambda query;
}

public class DataSpaceTemplateExecutablePointer extends DataSpaceTemplateExecutable 
{
    public PackageableElementPointer query;
}
````

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java [hunk 7]
```diff
@@ -282,6 +310,71 @@ else if (dataSpace.supportInfo instanceof DataSpaceSupportCombinedInfo)
...
+                                        if (templateExecutableQuery instanceof Root_meta_pure_metamodel_function_ConcreteFunctionDefinition_Impl)
+                                        {
+                                            Optional<? extends ValueSpecification> fromFunc = ((Root_meta_pure_metamodel_function_ConcreteFunctionDefinition_Impl<?>) templateExecutableQuery)._expressionSequence().toList().stream()
+                                                    .filter(func -> func instanceof Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl && ((Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl) func)._functionName.equals("from")).findAny();
+                                            if (fromFunc.isPresent())
+                                            {
+                                                Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl fromFuncExpression = (Root_meta_pure_metamodel_valuespecification_SimpleFunctionExpression_Impl) fromFunc.get();
+
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java:337]: Can we look into using pure built in functions for these types of checks to match things like function tests, so if we need to improve the logic of extracting runtime/mapping we can do it in one place ? 
See:
`Root_meta_pure_functions_meta_extractRuntimesFromFunctionDefinition_FunctionDefinition_1__Runtime_MANY_`

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/test/java/org/finos/legend/engine/language/pure/compiler/test/TestDataSpaceCompilationFromGrammar.java [hunk 8]
```diff
@@ -988,6 +988,143 @@ public void testDataSpaceWithTemplateExecutable()
...
+                "  executables:\n" +
+                "  [\n" +
+                "    {\n" +
+                "      id: 1;\n" +
+                "      title: 'Template 1';\n" +
+                "      query: model::templateFunc1():TabularDataSet[1];\n" +
+                "      executionContextKey: 'Context 1';\n" +
+                "    }\n" +
+                "  ];\n" +
+                "}\n", " at [44:1-65:1]: Error in 'model::dataSpace': The mapping utilized in the function within the curated template query does not align with the mapping applied in the execution context.");
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/test/java/org/finos/legend/engine/language/pure/compiler/test/TestDataSpaceCompilationFromGrammar.java:1099]: lets mention which execution context key it doesnt match with here.

**Review by MauricioUyaguari — CHANGES_REQUESTED** [review]

**Review by MauricioUyaguari — APPROVED** [review]

File: legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java [hunk 9]
```diff
@@ -282,6 +291,79 @@ else if (dataSpace.supportInfo instanceof DataSpaceSupportCombinedInfo)
...
+                                                    if (!mappingPath.equals(executionContext.mapping.path))
+                                                    {
+                                                        throw new EngineException("The mapping utilized in the function within the curated template query does not align with the mapping applied in the execution context `" + executionContext.name + "`.");
+                                                    }
+                                                }
+                                                // check if runtime matches to what is used in execution key
+                                                RichIterable<? extends org.finos.legend.pure.generated.Root_meta_core_runtime_Runtime> runtimes = org.finos.legend.pure.generated.core_pure_corefunctions_metaExtension.Root_meta_pure_functions_meta_extractRuntimesFromFunctionDefinition_FunctionDefinition_1__Runtime_MANY_(templateExecutableQuery, context.pureModel.getExecutionSupport());
```
Comment (MauricioUyaguari) [inline legend-engine-xts-data-space/legend-engine-xt-data-space-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/DataSpaceCompilerExtension.java:334]: Is there a similar function for mappings ? 
If not can we please build one out so its consistent with how we grab mappings

**Pull request merged**